# Machine Learning (ML): Google Cloud Platform (GCP): TensorFlow (TF): Financial Time-Series
# Model 01 = Binary Classifier
# Model 02 = Neural Network

## 1) Import Libraries

In [59]:
import StringIO

import pandas as pd
from pandas.tools.plotting import autocorrelation_plot
from pandas.tools.plotting import scatter_matrix

import numpy as np

import matplotlib.pyplot as plt

import gcp
import gcp.bigquery as bq

import tensorflow as tf

## 2) Install stuff

In [60]:
# Install xlrd
!pip install xlrd

Cleaning up...


## 3) Get the data from dropbox excel file

In [ ]:
# Save file on dropbox.com
# Right click to get the link
force = 1

# url_data_raw = 'https://www.dropbox.com/s/t8egr3lqxlpzcj8/data_raw.xlsm' # (dropbox for wylie.chan@gmail.com)
# url_data_raw = 'https://www.dropbox.com/s/9hlf1lrmqfy093v/data_raw.xlsm' #  (dropbox for eyup.saltik.2016@gmail.com)  
url_data_raw = 'https://www.dropbox.com/s/hwkuunzjvj9vph5/data_raw_03.xlsm' # updated 2016/07/04

In [ ]:
# Import file using url to file on dropbox
from urlparse import urlparse
from os.path import basename

bn_data_raw = basename(urlparse(url_data_raw).path)
print bn_data_raw

try:
    already_downloaded
except:
    already_downloaded = False
    
if force or not already_downloaded:
    already_downloaded = True
    !rm $bn_data_raw
    !wget $url_data_raw
#!ls -l

data_raw_03.xlsm
converted 'https://www.dropbox.com/s/hwkuunzjvj9vph5/data_raw_03.xlsm' (ANSI_X3.4-1968) -> 'https://www.dropbox.com/s/hwkuunzjvj9vph5/data_raw_03.xlsm' (UTF-8)
--2016-07-28 13:12:08--  https://www.dropbox.com/s/hwkuunzjvj9vph5/data_raw_03.xlsm
Resolving www.dropbox.com (www.dropbox.com)... 108.160.172.238
Connecting to www.dropbox.com (www.dropbox.com)|108.160.172.238|:443... connected.
HTTP request sent, awaiting response... 302 Found
Location: https://dl.dropboxusercontent.com/content_link/lv7q5Bf6h0hMJ79iS3Z1PPq2dnqhYlo0CFJTh3cRgIRf3NOeq841XRitKIi5dXJO/file [following]
converted 'https://dl.dropboxusercontent.com/content_link/lv7q5Bf6h0hMJ79iS3Z1PPq2dnqhYlo0CFJTh3cRgIRf3NOeq841XRitKIi5dXJO/file' (ANSI_X3.4-1968) -> 'https://dl.dropboxusercontent.com/content_link/lv7q5Bf6h0hMJ79iS3Z1PPq2dnqhYlo0CFJTh3cRgIRf3NOeq841XRitKIi5dXJO/file' (UTF-8)
--2016-07-28 13:12:09--  https://dl.dropboxusercontent.com/content_link/lv7q5Bf6h0hMJ79iS3Z1PPq2dnqhYlo0CFJTh3cRgIRf3NOeq841XRit

## 4) Read the Excel file raw data

In [ ]:
import numpy as np
import matplotlib.pyplot as plt
import pandas as pd
import xlrd

# Read the excel data
data_raw = pd.read_excel(bn_data_raw, 'data_raw')
# data = pd.read_excel(bn_data_raw, 'data_raw', header=0, index_col=0, parse_cols=None)
  # parse_cols=None: parse all columns
  # header=0:    sets the row 0 as col labels (ie. headers)
  # index_col=0: sets the col 0 as row labels (ie. index)
data_raw.head() # display the first few lines

## 5) Skip: Select only the columns that we have chosen from the PCA50 exercise

In [ ]:
# This list pertains to first set of data.
# list_cols_pca = [ 0,1,
#                   1070,788,926,112,69,574,654,1160,527,323,
#                   397,118,774,1028,1034,655,907,736,251,388,
#                   327,243,705,303,1146,467,136,1006,600,15,
#                   231,290,131,782,20,1048,630,1173,431,856,
#                   67,299,838,639,53,932,870,938,1061
#                 ]
# data_raw_pca = data_raw[list_cols_pca]

# For our new data set, we take all 86 columns
data_raw_pca = data_raw
data_raw_pca.head()

## 6) data_ticker_desc = Just a table to map column id to the ticker and name of the variable

In [ ]:
data_ticker_desc = data_raw_pca.ix[:1] # 1st 2 rows
data_ticker_desc

## 7) data_values = the actual values of the 86 variables

In [ ]:
data_values = data_raw_pca.ix[2:] # From row 2 onwards
data_values.head()

## 8) Set the dates as index of the dataframe

In [ ]:
data_values_indexed = data_values.set_index([0])
data_values_indexed.head()

In [ ]:
data_values_indexed.tail()

## 9) Fill any gaps in the data

In [ ]:
# Pandas includes a very convenient function for filling gaps in the data.
data_values_indexed = data_values_indexed.fillna(method='ffill')
data_values_indexed.head()

# Exploratory data analysis

Exploratory Data Analysis (EDA) is foundational to working with machine learning, and any other sort of analysis. EDA means getting to know your data, getting your fingers dirty with your data, feeling it and seeing it. The end result is you know your data very well, so when you build models you build them based on an actual, practical, physical understanding of the data, not assumptions or vaguely held notions. You can still make assumptions of course, but EDA means you will understand your assumptions and why you're making those assumptions. 

First, take a look at the data.

## 10) Describe the data briefly

In [ ]:
data_values_indexed.describe()

You can see that the various indices operate on scales differing by orders of magnitude. It's best to scale the data so that, for example, operations involving multiple indices aren't unduly influenced by a single, massive index.


## 11) Plot the data

In [ ]:
# N.B. A super-useful trick-ette is to assign the return value of plot to _ 
# so that you don't get text printed before the plot itself.

# _ = pd.concat([data_values_indexed[1],
#   data_values_indexed[1070],
#   data_values_indexed[788],
#   data_values_indexed[926]], axis=1).plot(figsize=(20, 15))

_ = data_values_indexed.plot(figsize=(20, 15))

As expected, the structure isn't uniformly visible for the indices. Divide each value in an individual index by the maximum value for that index., and then replot. The maximum value of all indices will be 1.

## 12) Calculate the max value for each column, prepare to scale data for each column

In [ ]:
data_values_indexed_max = data_values_indexed.max(axis=0) # max across axis 0 = rows
data_values_indexed_max.head()

## 13) Scale the data

In [ ]:
data_values_indexed_scaled = data_values_indexed / data_values_indexed_max
data_values_indexed_scaled.head()

## 14) Plot the scaled data

In [ ]:
_ = data_values_indexed_scaled.plot(figsize=(20, 15))

## 15) Auto-correlations

Next, plot autocorrelations for each of the indices. The autocorrelations determine correlations between current values of the index and lagged values of the same index. The goal is to determine whether the lagged values are reliable indicators of the current values. If they are, then we've identified a correlation.

In [ ]:
fig = plt.figure()
fig.set_figwidth(20)
fig.set_figheight(10)

_ = autocorrelation_plot(data_values_indexed_scaled)

For 2300 lagged days, we observe positive auto-correlations.
This suggests that as the variables increase, they tend to keep on increasing. Momentum.

After 2300 lagged days, we observe negative auto-correlations.

## 16) Skip: Just a reminder of the PCA columns we selected earlier

In [ ]:
# list_cols_pca = [ 0,1,
#                   1070,788,926,112,69,574,654,1160,527,323,
#                   397,118,774,1028,1034,655,907,736,251,388,
#                   327,243,705,303,1146,467,136,1006,600,15,
#                   231,290,131,782,20,1048,630,1173,431,856,
#                   67,299,838,639,53,932,870,938,1061
#                 ]

## 17) Scatter plots of the first 20 of our 86 variables vs USDMXN(varid=1)

In [ ]:
# _ = scatter_matrix(data_values_indexed_scaled)
dvis = data_values_indexed_scaled

_ = scatter_matrix(
      pd.concat(
      [ 
        dvis[ 1],dvis[ 2],dvis[ 3],dvis[ 4],dvis[ 5],dvis[ 6],dvis[ 7],dvis[ 8],dvis[ 9],dvis[10],
        dvis[11],dvis[12],dvis[13],dvis[14],dvis[15],dvis[16],dvis[17],dvis[18],dvis[19],dvis[20],
      ], axis=1), figsize=(15, 15), diagonal='kde')

## 18) Remind ourselves what our scaled data (of price or index LEVELS) looks like

In [ ]:
data_values_indexed_scaled.head()

## 19) Calculate Log Returns on our scaled data

In [ ]:
data_values_indexed_scaled_logret = pd.DataFrame()
data_values_indexed_scaled_logret = np.log(data_values_indexed_scaled/data_values_indexed_scaled.shift(-1)) # note dates are reverse-chrono

data_values_indexed_scaled_logret.head()

In [ ]:
data_values_indexed_scaled_logret.tail()

In [ ]:
data_values_indexed_scaled_logret.describe()

In [ ]:
# 2000/01/07 to 2016/05/27 = 1 header row + 4279 data rows
# print data_values_indexed_scaled_logret.head()
# print data_values_indexed_scaled_logret.tail()
print 'NumRowsIncludeHeader = len(data_values_indexed_scaled_logret) = ' + str(len(data_values_indexed_scaled_logret))

## 20) Replace inf, NaN in data with Zero

In [ ]:
data_values_indexed_scaled_logret = data_values_indexed_scaled_logret.replace([np.inf, -np.inf, np.nan], 0)
data_values_indexed_scaled_logret.describe()

## 21) Skip: Fill the Gaps

In [ ]:
# Pandas includes a very convenient function for filling gaps in the data.
# data_values_indexed_scaled_logret = data_values_indexed_scaled_logret.fillna(method='ffill')
data_values_indexed_scaled_logret.head()

## 22) Plot log returns of scaled data

In [ ]:
_ = data_values_indexed_scaled_logret.plot(figsize=(20, 15))

## 23) Auto-Correlations of log returns of scaled data

In [ ]:
fig = plt.figure()
fig.set_figwidth(20)
fig.set_figheight(10)

_ = autocorrelation_plot(data_values_indexed_scaled_logret)

There are no auto-correlations, so we are good.

## 24) Skip: Just a reminder of the PCA50 Columns we selected earlier

In [ ]:
# list_cols_pca = [ 0,1,
#                   1070,788,926,112,69,574,654,1160,527,323,
#                   397,118,774,1028,1034,655,907,736,251,388,
#                   327,243,705,303,1146,467,136,1006,600,15,
#                   231,290,131,782,20,1048,630,1173,431,856,
#                   67,299,838,639,53,932,870,938,1061
#                 ]

## 25) Scatter plots of log returns of first 20 of the 86 variables where log-returns of USDMXN(varid=1)

In [ ]:
# _ = scatter_matrix(data_values_indexed_scaled_logret)   # takes long time, becareful, save work first
dvislr = data_values_indexed_scaled_logret

_ = scatter_matrix(
      pd.concat(
      [ 
        dvislr[ 1],dvislr[ 2],dvislr[ 3],dvislr[ 4],dvislr[ 5],dvislr[ 6],dvislr[ 7],dvislr[ 8],dvislr[ 9],dvislr[10],
        dvislr[11],dvislr[12],dvislr[13],dvislr[14],dvislr[15],dvislr[16],dvislr[17],dvislr[18],dvislr[19],dvislr[20],
      ], axis=1), figsize=(15, 15), diagonal='kde')


In [ ]:
data_values_indexed_scaled_logret.head()

## Summing up the EDA

At this point, you've done a good enough job of exploratory data analysis. You've visualized our data and come to know it better. 
You've transformed it into a form that is useful for modelling, log returns, and looked at how indices relate to each other. 

What should we think so far?

Cloud Datalab is working great. With just a few lines of code, you were able to munge the data, visualize the changes, and make decisions. You could easily analyze and iterate. This is a common feature of iPython, but the advantage here is that Cloud Datalab is a managed service that you can simply click and use, so you can focus on your analysis.

## Feature selection

At this point, we can see a model:

* We'll predict whether the USDMXN close today will be higher or lower than yesterday.

Predicting whether the log return of the USDMXN is positive or negative is a classification problem. 
That is, we want to choose one option from a finite set of options, in this case positive or negative. 
This is the base case of classification where we have only two values to choose from, known as binary classification, or logistic regression.

Machine learning models are very good at finding weak signals from data.
In machine learning, as in most things, there are subtle tradeoffs happening, but in general good data is better than good algorithms, which are better than good frameworks. 
You need all three pillars but in that order of importance: data, algorithms, frameworks.

## TensorFlow

[TensorFlow](https://tensorflow.org) is an open source software library, initiated by Google, for numerical computation using data flow graphs. TensorFlow is based on Google's machine learning expertise and is the next generation framework used internally at Google for tasks such as translation and image recognition. It's a wonderful framework for machine learning because it's expressive, efficient, and easy to use.

### Feature engineering for TensorFlow

From a training and testing perspective, time series data is easy. Training data should come from events that happened before test data events, and be contiguous in time.  Otherwise,  your model would be trained on events from "the future", at least as compared to the test data. It would then likely perform badly in practice, because you can’t really have access to data from the future. That means random sampling or cross validation don't apply to time series data. Decide on a training-versus-testing split, and divide your data into training and test datasets.

In this case, you'll create the features together with two additional columns:

* usdmxn_logret_positive, which is 1 if the log return of the USDMXN close is positive, and 0 otherwise. 
* usdmxn_logret_negative, which is 1 if the log return of the USDMXN close is negative, and 1 otherwise. 

We'll use 80% of our data for training and 20% for testing.

# Model 1: Binary Classification (BC) = Logistic Regression (LR)

## Model 1: Step 01: Indicator Columns

In [ ]:
# 00: Original: Initialize indicator columns to 0
data_values_indexed_scaled_logret['usdmxn_logret_positive'] = 0
data_values_indexed_scaled_logret['usdmxn_logret_negative'] = 0

In [ ]:
# 00: Populate results columns according to actual usdmxn returns (positive or negative)
data_values_indexed_scaled_logret.ix[data_values_indexed_scaled_logret[1] >= 0, 'usdmxn_logret_positive'] = 1
data_values_indexed_scaled_logret.ix[data_values_indexed_scaled_logret[1] <  0, 'usdmxn_logret_negative'] = 1

In [ ]:
# 01: Noise Cancellation
data_values_indexed_scaled_logret['usdmxn_logret_positive_0050bp'] = 0
data_values_indexed_scaled_logret['usdmxn_logret_negative_0050bp'] = 0

In [ ]:
# 01: Populate results columns according to actual usdmxn returns (outside noise zone above or below)
data_values_indexed_scaled_logret.ix[data_values_indexed_scaled_logret[1] >= 0.0050, 'usdmxn_logret_positive_0050bp'] = 1
data_values_indexed_scaled_logret.ix[data_values_indexed_scaled_logret[1] <  -0.0050, 'usdmxn_logret_negative_0050bp'] = 1

In [ ]:
# # 02: Stratification
# data_values_indexed_scaled_logret['usdmxn_logret_+0150bp_and_above'] = 0; data_values_indexed_scaled_logret['usdmxn_logret_+0150bp_and_below'] = 0;
# data_values_indexed_scaled_logret['usdmxn_logret_+0140bp_and_above'] = 0; data_values_indexed_scaled_logret['usdmxn_logret_+0140bp_and_below'] = 0;
# data_values_indexed_scaled_logret['usdmxn_logret_+0130bp_and_above'] = 0; data_values_indexed_scaled_logret['usdmxn_logret_+0130bp_and_below'] = 0;
# data_values_indexed_scaled_logret['usdmxn_logret_+0120bp_and_above'] = 0; data_values_indexed_scaled_logret['usdmxn_logret_+0120bp_and_below'] = 0;
# data_values_indexed_scaled_logret['usdmxn_logret_+0110bp_and_above'] = 0; data_values_indexed_scaled_logret['usdmxn_logret_+0110bp_and_below'] = 0;
# data_values_indexed_scaled_logret['usdmxn_logret_+0100bp_and_above'] = 0; data_values_indexed_scaled_logret['usdmxn_logret_+0100bp_and_below'] = 0;
# data_values_indexed_scaled_logret['usdmxn_logret_+0090bp_and_above'] = 0; data_values_indexed_scaled_logret['usdmxn_logret_+0090bp_and_below'] = 0;
# data_values_indexed_scaled_logret['usdmxn_logret_+0080bp_and_above'] = 0; data_values_indexed_scaled_logret['usdmxn_logret_+0080bp_and_below'] = 0;
# data_values_indexed_scaled_logret['usdmxn_logret_+0070bp_and_above'] = 0; data_values_indexed_scaled_logret['usdmxn_logret_+0070bp_and_below'] = 0;
# data_values_indexed_scaled_logret['usdmxn_logret_+0060bp_and_above'] = 0; data_values_indexed_scaled_logret['usdmxn_logret_+0060bp_and_below'] = 0;
# data_values_indexed_scaled_logret['usdmxn_logret_+0050bp_and_above'] = 0; data_values_indexed_scaled_logret['usdmxn_logret_+0050bp_and_below'] = 0;
# data_values_indexed_scaled_logret['usdmxn_logret_+0040bp_and_above'] = 0; data_values_indexed_scaled_logret['usdmxn_logret_+0040bp_and_below'] = 0;
# data_values_indexed_scaled_logret['usdmxn_logret_+0030bp_and_above'] = 0; data_values_indexed_scaled_logret['usdmxn_logret_+0030bp_and_below'] = 0;
# data_values_indexed_scaled_logret['usdmxn_logret_+0020bp_and_above'] = 0; data_values_indexed_scaled_logret['usdmxn_logret_+0020bp_and_below'] = 0;
# data_values_indexed_scaled_logret['usdmxn_logret_+0010bp_and_above'] = 0; data_values_indexed_scaled_logret['usdmxn_logret_+0010bp_and_below'] = 0;
# data_values_indexed_scaled_logret['usdmxn_logret_+0000bp_and_above'] = 0; data_values_indexed_scaled_logret['usdmxn_logret_+0000bp_and_below'] = 0;
# data_values_indexed_scaled_logret['usdmxn_logret_-0010bp_and_above'] = 0; data_values_indexed_scaled_logret['usdmxn_logret_-0010bp_and_below'] = 0;
# data_values_indexed_scaled_logret['usdmxn_logret_-0020bp_and_above'] = 0; data_values_indexed_scaled_logret['usdmxn_logret_-0020bp_and_below'] = 0;
# data_values_indexed_scaled_logret['usdmxn_logret_-0030bp_and_above'] = 0; data_values_indexed_scaled_logret['usdmxn_logret_-0030bp_and_below'] = 0;
# data_values_indexed_scaled_logret['usdmxn_logret_-0040bp_and_above'] = 0; data_values_indexed_scaled_logret['usdmxn_logret_-0040bp_and_below'] = 0;
# data_values_indexed_scaled_logret['usdmxn_logret_-0050bp_and_above'] = 0; data_values_indexed_scaled_logret['usdmxn_logret_-0050bp_and_below'] = 0;
# data_values_indexed_scaled_logret['usdmxn_logret_-0060bp_and_above'] = 0; data_values_indexed_scaled_logret['usdmxn_logret_-0060bp_and_below'] = 0;
# data_values_indexed_scaled_logret['usdmxn_logret_-0070bp_and_above'] = 0; data_values_indexed_scaled_logret['usdmxn_logret_-0070bp_and_below'] = 0;
# data_values_indexed_scaled_logret['usdmxn_logret_-0080bp_and_above'] = 0; data_values_indexed_scaled_logret['usdmxn_logret_-0080bp_and_below'] = 0;
# data_values_indexed_scaled_logret['usdmxn_logret_-0090bp_and_above'] = 0; data_values_indexed_scaled_logret['usdmxn_logret_-0090bp_and_below'] = 0;
# data_values_indexed_scaled_logret['usdmxn_logret_-0100bp_and_above'] = 0; data_values_indexed_scaled_logret['usdmxn_logret_-0100bp_and_below'] = 0;
# data_values_indexed_scaled_logret['usdmxn_logret_-0110bp_and_above'] = 0; data_values_indexed_scaled_logret['usdmxn_logret_-0110bp_and_below'] = 0;
# data_values_indexed_scaled_logret['usdmxn_logret_-0120bp_and_above'] = 0; data_values_indexed_scaled_logret['usdmxn_logret_-0120bp_and_below'] = 0;
# data_values_indexed_scaled_logret['usdmxn_logret_-0130bp_and_above'] = 0; data_values_indexed_scaled_logret['usdmxn_logret_-0130bp_and_below'] = 0;
# data_values_indexed_scaled_logret['usdmxn_logret_-0140bp_and_above'] = 0; data_values_indexed_scaled_logret['usdmxn_logret_-0140bp_and_below'] = 0;
# data_values_indexed_scaled_logret['usdmxn_logret_-0150bp_and_above'] = 0; data_values_indexed_scaled_logret['usdmxn_logret_-0150bp_and_below'] = 0;

In [ ]:
# 02: Stratification: Initialize
for d_bp in xrange(-150,160,10): # -150:10:+150
  d = d_bp / 10000.00
  s_above = 'usdmxn_logret_' + '{0:+05d}'.format(d_bp) + 'bp_and_above'
  s_below = 'usdmxn_logret_' + '{0:+05d}'.format(d_bp) + 'bp_and_below'
  # print d
  # print s_above
  # print s_below
  data_values_indexed_scaled_logret[s_above] = 0;
  data_values_indexed_scaled_logret[s_below] = 0;

In [ ]:
# 02: Stratification: Populate results columns according to actual usdmxn returns (above or below each level)
for d_bp in xrange(-150,160,10): # -150:10:+150
  d = d_bp / 10000.00
  s_above = 'usdmxn_logret_' + '{0:+05d}'.format(d_bp) + 'bp_and_above'
  s_below = 'usdmxn_logret_' + '{0:+05d}'.format(d_bp) + 'bp_and_below'
  # print d
  # print s_above
  # print s_below
  data_values_indexed_scaled_logret.ix[data_values_indexed_scaled_logret[1] >= d, s_above] = 1
  data_values_indexed_scaled_logret.ix[data_values_indexed_scaled_logret[1] <  d, s_below] = 1


In [ ]:
data_values_indexed_scaled_logret.head()

In [ ]:
data_values_indexed_scaled_logret.describe()

## Model 1: Step 01.01: Remove data noise in a stratified way

In [ ]:
dvislr = data_values_indexed_scaled_logret
# dvislrnr = data_values_indexed_scaled_logret (noise removed)

# Example Filter: df[(df.A == 1) & (df.D == 6)]
dvislr_nr_m0010_z0000_p0010 = dvislr[ (dvislr['usdmxn_logret_-0010bp_and_below'] == 1) | (dvislr['usdmxn_logret_+0010bp_and_above'] == 1) ] # remove data between [-0.10%, +0.10%], center at +0.00%
dvislr_nr_m0020_z0000_p0020 = dvislr[ (dvislr['usdmxn_logret_-0020bp_and_below'] == 1) | (dvislr['usdmxn_logret_+0020bp_and_above'] == 1) ] # remove data between [-0.10%, +0.10%], center at +0.00%
dvislr_nr_m0030_z0000_p0030 = dvislr[ (dvislr['usdmxn_logret_-0030bp_and_below'] == 1) | (dvislr['usdmxn_logret_+0030bp_and_above'] == 1) ] # remove data between [-0.10%, +0.10%], center at +0.00%
dvislr_nr_m0040_z0000_p0040 = dvislr[ (dvislr['usdmxn_logret_-0040bp_and_below'] == 1) | (dvislr['usdmxn_logret_+0040bp_and_above'] == 1) ] # remove data between [-0.10%, +0.10%], center at +0.00%
dvislr_nr_m0050_z0000_p0050 = dvislr[ (dvislr['usdmxn_logret_-0050bp_and_below'] == 1) | (dvislr['usdmxn_logret_+0050bp_and_above'] == 1) ] # remove data between [-0.50%, +0.50%], center at +0.00%
dvislr_nr_m0060_z0000_p0060 = dvislr[ (dvislr['usdmxn_logret_-0060bp_and_below'] == 1) | (dvislr['usdmxn_logret_+0060bp_and_above'] == 1) ] # remove data between [-0.50%, +0.50%], center at +0.00%
dvislr_nr_m0070_z0000_p0070 = dvislr[ (dvislr['usdmxn_logret_-0070bp_and_below'] == 1) | (dvislr['usdmxn_logret_+0070bp_and_above'] == 1) ] # remove data between [-0.50%, +0.50%], center at +0.00%
dvislr_nr_m0080_z0000_p0080 = dvislr[ (dvislr['usdmxn_logret_-0080bp_and_below'] == 1) | (dvislr['usdmxn_logret_+0080bp_and_above'] == 1) ] # remove data between [-0.50%, +0.50%], center at +0.00%
dvislr_nr_m0090_z0000_p0090 = dvislr[ (dvislr['usdmxn_logret_-0090bp_and_below'] == 1) | (dvislr['usdmxn_logret_+0090bp_and_above'] == 1) ] # remove data between [-0.50%, +0.50%], center at +0.00%
dvislr_nr_m0100_z0000_p0100 = dvislr[ (dvislr['usdmxn_logret_-0100bp_and_below'] == 1) | (dvislr['usdmxn_logret_+0100bp_and_above'] == 1) ] # remove data between [-0.50%, +0.50%], center at +0.00%

# check rows
print 'dvislr_nr_m0010_z0000_p0010 rows =' + str(len(dvislr_nr_m0010_z0000_p0010))
print 'dvislr_nr_m0020_z0000_p0020 rows =' + str(len(dvislr_nr_m0020_z0000_p0020))
print 'dvislr_nr_m0030_z0000_p0030 rows =' + str(len(dvislr_nr_m0030_z0000_p0030))
print 'dvislr_nr_m0040_z0000_p0040 rows =' + str(len(dvislr_nr_m0040_z0000_p0040))
print 'dvislr_nr_m0050_z0000_p0050 rows =' + str(len(dvislr_nr_m0050_z0000_p0050))
print 'dvislr_nr_m0060_z0000_p0060 rows =' + str(len(dvislr_nr_m0060_z0000_p0060))
print 'dvislr_nr_m0070_z0000_p0070 rows =' + str(len(dvislr_nr_m0070_z0000_p0070))
print 'dvislr_nr_m0080_z0000_p0080 rows =' + str(len(dvislr_nr_m0080_z0000_p0080))
print 'dvislr_nr_m0090_z0000_p0090 rows =' + str(len(dvislr_nr_m0090_z0000_p0090))
print 'dvislr_nr_m0100_z0000_p0100 rows =' + str(len(dvislr_nr_m0100_z0000_p0100))

# check rows results
# dvislr_nr_m0010_z0000_p0010 rows =3554
# dvislr_nr_m0020_z0000_p0020 rows =2904
# dvislr_nr_m0030_z0000_p0030 rows =2316
# dvislr_nr_m0040_z0000_p0040 rows =1837
# dvislr_nr_m0050_z0000_p0050 rows =1437
# dvislr_nr_m0060_z0000_p0060 rows =1135
# dvislr_nr_m0070_z0000_p0070 rows =893
# dvislr_nr_m0080_z0000_p0080 rows =697
# dvislr_nr_m0090_z0000_p0090 rows =559
# dvislr_nr_m0100_z0000_p0100 rows =439

## Model 1: Step 02: Split data into training, test

### training test data empty shell

In [ ]:
# training_test_data
# col 01-02 = 02 cols = indicators of results
# col 03-88 = 86 cols = inputs
training_test_data = pd.DataFrame(
  columns=[
    # 'usdmxn_logret_positive', 'usdmxn_logret_negative',
    # 'usdmxn_logret_positive_0050bp', 'usdmxn_logret_negative_0050bp',

    'bc_up', 'bc_dn',
    
    'usdmxn_logret_-0150bp_and_above', 'usdmxn_logret_-0150bp_and_below',
    'usdmxn_logret_-0140bp_and_above', 'usdmxn_logret_-0140bp_and_below',
    'usdmxn_logret_-0130bp_and_above', 'usdmxn_logret_-0130bp_and_below',
    'usdmxn_logret_-0120bp_and_above', 'usdmxn_logret_-0120bp_and_below',
    'usdmxn_logret_-0110bp_and_above', 'usdmxn_logret_-0110bp_and_below',
    'usdmxn_logret_-0100bp_and_above', 'usdmxn_logret_-0100bp_and_below',
    'usdmxn_logret_-0090bp_and_above', 'usdmxn_logret_-0090bp_and_below',
    'usdmxn_logret_-0080bp_and_above', 'usdmxn_logret_-0080bp_and_below',
    'usdmxn_logret_-0070bp_and_above', 'usdmxn_logret_-0070bp_and_below',
    'usdmxn_logret_-0060bp_and_above', 'usdmxn_logret_-0060bp_and_below',
    'usdmxn_logret_-0050bp_and_above', 'usdmxn_logret_-0050bp_and_below',
    'usdmxn_logret_-0040bp_and_above', 'usdmxn_logret_-0040bp_and_below',
    'usdmxn_logret_-0030bp_and_above', 'usdmxn_logret_-0030bp_and_below',
    'usdmxn_logret_-0020bp_and_above', 'usdmxn_logret_-0020bp_and_below',
    'usdmxn_logret_-0010bp_and_above', 'usdmxn_logret_-0010bp_and_below',
    'usdmxn_logret_+0000bp_and_above', 'usdmxn_logret_+0000bp_and_below',
    'usdmxn_logret_+0010bp_and_above', 'usdmxn_logret_+0010bp_and_below',
    'usdmxn_logret_+0020bp_and_above', 'usdmxn_logret_+0020bp_and_below',
    'usdmxn_logret_+0030bp_and_above', 'usdmxn_logret_+0030bp_and_below',
    'usdmxn_logret_+0040bp_and_above', 'usdmxn_logret_+0040bp_and_below',
    'usdmxn_logret_+0050bp_and_above', 'usdmxn_logret_+0050bp_and_below',
    'usdmxn_logret_+0060bp_and_above', 'usdmxn_logret_+0060bp_and_below',
    'usdmxn_logret_+0070bp_and_above', 'usdmxn_logret_+0070bp_and_below',
    'usdmxn_logret_+0080bp_and_above', 'usdmxn_logret_+0080bp_and_below',
    'usdmxn_logret_+0090bp_and_above', 'usdmxn_logret_+0090bp_and_below',
    'usdmxn_logret_+0100bp_and_above', 'usdmxn_logret_+0100bp_and_below',
    'usdmxn_logret_+0110bp_and_above', 'usdmxn_logret_+0110bp_and_below',
    'usdmxn_logret_+0120bp_and_above', 'usdmxn_logret_+0120bp_and_below',
    'usdmxn_logret_+0130bp_and_above', 'usdmxn_logret_+0130bp_and_below',
    'usdmxn_logret_+0140bp_and_above', 'usdmxn_logret_+0140bp_and_below',
    'usdmxn_logret_+0150bp_and_above', 'usdmxn_logret_+0150bp_and_below',

    '1','2','3','4','5','6','7','8','9','10',
    '11','12','13','14','15','16','17','18','19','20',
    '21','22','23','24','25','26','27','28','29','30',
    '31','32','33','34','35','36','37','38','39','40',
    '41','42','43','44','45','46','47','48','49','50',
    '51','52','53','54','55','56','57','58','59','60',
    '61','62','63','64','65','66','67','68','69','70',
    '71','72','73','74','75','76','77','78','79','80',
    '81','82','83','84','85','86',
  ])

training_test_data.head()

### Read data into variables

In [ ]:
# row 0      = header
# row 1-4279 = data
# NumRowsIncludeHeader = len(data_values_indexed_scaled_logret) = 4280 = 0-4279
# Start from row 7, so we can have up to (7-1) lookback days

# check rows results
# dvislr_nr_m0010_z0000_p0010 rows =3554
# dvislr_nr_m0020_z0000_p0020 rows =2904
# dvislr_nr_m0030_z0000_p0030 rows =2316
# dvislr_nr_m0040_z0000_p0040 rows =1837
# dvislr_nr_m0050_z0000_p0050 rows =1437
# dvislr_nr_m0060_z0000_p0060 rows =1135
# dvislr_nr_m0070_z0000_p0070 rows =893
# dvislr_nr_m0080_z0000_p0080 rows =697
# dvislr_nr_m0090_z0000_p0090 rows =559
# dvislr_nr_m0100_z0000_p0100 rows =439

#==========================================================================================
# Select the data
#==========================================================================================
# dvislr =  dvislr_nr_m0010_z0000_p0010
# dvislr =  dvislr_nr_m0020_z0000_p0020
# dvislr =  dvislr_nr_m0030_z0000_p0030
# dvislr =  dvislr_nr_m0040_z0000_p0040
# dvislr =  dvislr_nr_m0050_z0000_p0050
# dvislr =  dvislr_nr_m0060_z0000_p0060
dvislr =  dvislr_nr_m0070_z0000_p0070
# dvislr =  dvislr_nr_m0080_z0000_p0080
# dvislr =  dvislr_nr_m0090_z0000_p0090
# dvislr =  dvislr_nr_m0100_z0000_p0100

for i in range(7, len(dvislr)): # [7, 4280) = [7,4279] 

  # 00: original
  #   usdmxn_logret_positive = data_values_indexed_scaled_logret['usdmxn_logret_positive'].ix[i]
  #   usdmxn_logret_negative = data_values_indexed_scaled_logret['usdmxn_logret_negative'].ix[i]

  # 01: noise cancellation
  #   usdmxn_logret_positive_0050bp = data_values_indexed_scaled_logret['usdmxn_logret_positive_0050bp'].ix[i]
  #   usdmxn_logret_negative_0050bp = data_values_indexed_scaled_logret['usdmxn_logret_negative_0050bp'].ix[i]

  # 02: stratification  
  usdmxn_logret_p0150bp_and_below = dvislr['usdmxn_logret_+0150bp_and_below'].ix[i]; usdmxn_logret_p0150bp_and_above = dvislr['usdmxn_logret_+0150bp_and_above'].ix[i];
  usdmxn_logret_p0140bp_and_below = dvislr['usdmxn_logret_+0140bp_and_below'].ix[i]; usdmxn_logret_p0140bp_and_above = dvislr['usdmxn_logret_+0140bp_and_above'].ix[i];
  usdmxn_logret_p0130bp_and_below = dvislr['usdmxn_logret_+0130bp_and_below'].ix[i]; usdmxn_logret_p0130bp_and_above = dvislr['usdmxn_logret_+0130bp_and_above'].ix[i];
  usdmxn_logret_p0120bp_and_below = dvislr['usdmxn_logret_+0120bp_and_below'].ix[i]; usdmxn_logret_p0120bp_and_above = dvislr['usdmxn_logret_+0120bp_and_above'].ix[i];
  usdmxn_logret_p0110bp_and_below = dvislr['usdmxn_logret_+0110bp_and_below'].ix[i]; usdmxn_logret_p0110bp_and_above = dvislr['usdmxn_logret_+0110bp_and_above'].ix[i];
  
  usdmxn_logret_p0100bp_and_below = dvislr['usdmxn_logret_+0100bp_and_below'].ix[i]; usdmxn_logret_p0100bp_and_above = dvislr['usdmxn_logret_+0100bp_and_above'].ix[i];
  usdmxn_logret_p0090bp_and_below = dvislr['usdmxn_logret_+0090bp_and_below'].ix[i]; usdmxn_logret_p0090bp_and_above = dvislr['usdmxn_logret_+0090bp_and_above'].ix[i];
  usdmxn_logret_p0080bp_and_below = dvislr['usdmxn_logret_+0080bp_and_below'].ix[i]; usdmxn_logret_p0080bp_and_above = dvislr['usdmxn_logret_+0080bp_and_above'].ix[i];
  usdmxn_logret_p0070bp_and_below = dvislr['usdmxn_logret_+0070bp_and_below'].ix[i]; usdmxn_logret_p0070bp_and_above = dvislr['usdmxn_logret_+0070bp_and_above'].ix[i];
  usdmxn_logret_p0060bp_and_below = dvislr['usdmxn_logret_+0060bp_and_below'].ix[i]; usdmxn_logret_p0060bp_and_above = dvislr['usdmxn_logret_+0060bp_and_above'].ix[i];

  usdmxn_logret_p0050bp_and_below = dvislr['usdmxn_logret_+0050bp_and_below'].ix[i]; usdmxn_logret_p0050bp_and_above = dvislr['usdmxn_logret_+0050bp_and_above'].ix[i];
  usdmxn_logret_p0040bp_and_below = dvislr['usdmxn_logret_+0040bp_and_below'].ix[i]; usdmxn_logret_p0040bp_and_above = dvislr['usdmxn_logret_+0040bp_and_above'].ix[i];
  usdmxn_logret_p0030bp_and_below = dvislr['usdmxn_logret_+0030bp_and_below'].ix[i]; usdmxn_logret_p0030bp_and_above = dvislr['usdmxn_logret_+0030bp_and_above'].ix[i];
  usdmxn_logret_p0020bp_and_below = dvislr['usdmxn_logret_+0020bp_and_below'].ix[i]; usdmxn_logret_p0020bp_and_above = dvislr['usdmxn_logret_+0020bp_and_above'].ix[i];
  usdmxn_logret_p0010bp_and_below = dvislr['usdmxn_logret_+0010bp_and_below'].ix[i]; usdmxn_logret_p0010bp_and_above = dvislr['usdmxn_logret_+0010bp_and_above'].ix[i];

  usdmxn_logret_p0000bp_and_below = dvislr['usdmxn_logret_+0000bp_and_below'].ix[i]; usdmxn_logret_p0000bp_and_above = dvislr['usdmxn_logret_+0000bp_and_above'].ix[i];

  usdmxn_logret_m0010bp_and_below = dvislr['usdmxn_logret_-0010bp_and_below'].ix[i]; usdmxn_logret_m0010bp_and_above = dvislr['usdmxn_logret_-0010bp_and_above'].ix[i];
  usdmxn_logret_m0020bp_and_below = dvislr['usdmxn_logret_-0020bp_and_below'].ix[i]; usdmxn_logret_m0020bp_and_above = dvislr['usdmxn_logret_-0020bp_and_above'].ix[i];
  usdmxn_logret_m0030bp_and_below = dvislr['usdmxn_logret_-0030bp_and_below'].ix[i]; usdmxn_logret_m0030bp_and_above = dvislr['usdmxn_logret_-0030bp_and_above'].ix[i];
  usdmxn_logret_m0040bp_and_below = dvislr['usdmxn_logret_-0040bp_and_below'].ix[i]; usdmxn_logret_m0040bp_and_above = dvislr['usdmxn_logret_-0040bp_and_above'].ix[i];
  usdmxn_logret_m0050bp_and_below = dvislr['usdmxn_logret_-0050bp_and_below'].ix[i]; usdmxn_logret_m0050bp_and_above = dvislr['usdmxn_logret_-0050bp_and_above'].ix[i];

  usdmxn_logret_m0060bp_and_below = dvislr['usdmxn_logret_-0060bp_and_below'].ix[i]; usdmxn_logret_m0060bp_and_above = dvislr['usdmxn_logret_-0060bp_and_above'].ix[i];
  usdmxn_logret_m0070bp_and_below = dvislr['usdmxn_logret_-0070bp_and_below'].ix[i]; usdmxn_logret_m0070bp_and_above = dvislr['usdmxn_logret_-0070bp_and_above'].ix[i];
  usdmxn_logret_m0080bp_and_below = dvislr['usdmxn_logret_-0080bp_and_below'].ix[i]; usdmxn_logret_m0080bp_and_above = dvislr['usdmxn_logret_-0080bp_and_above'].ix[i];
  usdmxn_logret_m0090bp_and_below = dvislr['usdmxn_logret_-0090bp_and_below'].ix[i]; usdmxn_logret_m0090bp_and_above = dvislr['usdmxn_logret_-0090bp_and_above'].ix[i];
  usdmxn_logret_m0100bp_and_below = dvislr['usdmxn_logret_-0100bp_and_below'].ix[i]; usdmxn_logret_m0100bp_and_above = dvislr['usdmxn_logret_-0100bp_and_above'].ix[i];

  usdmxn_logret_m0110bp_and_below = dvislr['usdmxn_logret_-0110bp_and_below'].ix[i]; usdmxn_logret_m0110bp_and_above = dvislr['usdmxn_logret_-0110bp_and_above'].ix[i];
  usdmxn_logret_m0120bp_and_below = dvislr['usdmxn_logret_-0120bp_and_below'].ix[i]; usdmxn_logret_m0120bp_and_above = dvislr['usdmxn_logret_-0120bp_and_above'].ix[i];
  usdmxn_logret_m0130bp_and_below = dvislr['usdmxn_logret_-0130bp_and_below'].ix[i]; usdmxn_logret_m0130bp_and_above = dvislr['usdmxn_logret_-0130bp_and_above'].ix[i];
  usdmxn_logret_m0140bp_and_below = dvislr['usdmxn_logret_-0140bp_and_below'].ix[i]; usdmxn_logret_m0140bp_and_above = dvislr['usdmxn_logret_-0140bp_and_above'].ix[i];
  usdmxn_logret_m0150bp_and_below = dvislr['usdmxn_logret_-0150bp_and_below'].ix[i]; usdmxn_logret_m0150bp_and_above = dvislr['usdmxn_logret_-0150bp_and_above'].ix[i];

  #==========================================================================================
  # Select 2 levels: up and dn
  #==========================================================================================
#   bc_dn = usdmxn_logret_m0010bp_and_below; bc_up = usdmxn_logret_p0010bp_and_above;
#   bc_dn = usdmxn_logret_m0020bp_and_below; bc_up = usdmxn_logret_p0020bp_and_above;
#   bc_dn = usdmxn_logret_m0030bp_and_below; bc_up = usdmxn_logret_p0030bp_and_above;
#   bc_dn = usdmxn_logret_m0040bp_and_below; bc_up = usdmxn_logret_p0040bp_and_above;
#   bc_dn = usdmxn_logret_m0050bp_and_below; bc_up = usdmxn_logret_p0050bp_and_above;
#   bc_dn = usdmxn_logret_m0060bp_and_below; bc_up = usdmxn_logret_p0060bp_and_above;
  bc_dn = usdmxn_logret_m0070bp_and_below; bc_up = usdmxn_logret_p0070bp_and_above;
#   bc_dn = usdmxn_logret_m0080bp_and_below; bc_up = usdmxn_logret_p0080bp_and_above;
#   bc_dn = usdmxn_logret_m0090bp_and_below; bc_up = usdmxn_logret_p0090bp_and_above;
#   bc_dn = usdmxn_logret_m0100bp_and_below; bc_up = usdmxn_logret_p0100bp_and_above;
  
  #   v_001_Tm005 = data_values_indexed_scaled_logret['1'].ix[i-5]  # lookback 1d

  v_001_Tm000 = dvislr[1].ix[i];  v_002_Tm000 = dvislr[2].ix[i];  v_003_Tm000 = dvislr[3].ix[i];  v_004_Tm000 = dvislr[4].ix[i];  v_005_Tm000 = dvislr[5].ix[i];
  v_006_Tm000 = dvislr[6].ix[i];  v_007_Tm000 = dvislr[7].ix[i];  v_008_Tm000 = dvislr[8].ix[i];  v_009_Tm000 = dvislr[9].ix[i];  v_010_Tm000 = dvislr[10].ix[i]

  v_011_Tm000 = dvislr[11].ix[i]; v_012_Tm000 = dvislr[12].ix[i]; v_013_Tm000 = dvislr[13].ix[i]; v_014_Tm000 = dvislr[14].ix[i]; v_015_Tm000 = dvislr[15].ix[i];
  v_016_Tm000 = dvislr[16].ix[i]; v_017_Tm000 = dvislr[17].ix[i]; v_018_Tm000 = dvislr[18].ix[i]; v_019_Tm000 = dvislr[19].ix[i]; v_020_Tm000 = dvislr[20].ix[i];
  
  v_021_Tm000 = dvislr[21].ix[i]; v_022_Tm000 = dvislr[22].ix[i]; v_023_Tm000 = dvislr[23].ix[i]; v_024_Tm000 = dvislr[24].ix[i]; v_025_Tm000 = dvislr[25].ix[i];
  v_026_Tm000 = dvislr[26].ix[i]; v_027_Tm000 = dvislr[27].ix[i]; v_028_Tm000 = dvislr[28].ix[i]; v_029_Tm000 = dvislr[29].ix[i]; v_030_Tm000 = dvislr[30].ix[i];

  v_031_Tm000 = dvislr[31].ix[i]; v_032_Tm000 = dvislr[32].ix[i]; v_033_Tm000 = dvislr[33].ix[i]; v_034_Tm000 = dvislr[34].ix[i]; v_035_Tm000 = dvislr[35].ix[i];
  v_036_Tm000 = dvislr[36].ix[i]; v_037_Tm000 = dvislr[37].ix[i]; v_038_Tm000 = dvislr[38].ix[i]; v_039_Tm000 = dvislr[39].ix[i]; v_040_Tm000 = dvislr[40].ix[i];

  v_041_Tm000 = dvislr[41].ix[i]; v_042_Tm000 = dvislr[42].ix[i]; v_043_Tm000 = dvislr[43].ix[i]; v_044_Tm000 = dvislr[44].ix[i]; v_045_Tm000 = dvislr[45].ix[i];
  v_046_Tm000 = dvislr[46].ix[i]; v_047_Tm000 = dvislr[47].ix[i]; v_048_Tm000 = dvislr[48].ix[i]; v_049_Tm000 = dvislr[49].ix[i]; v_050_Tm000 = dvislr[50].ix[i];

  v_051_Tm000 = dvislr[51].ix[i]; v_052_Tm000 = dvislr[52].ix[i]; v_053_Tm000 = dvislr[53].ix[i]; v_054_Tm000 = dvislr[54].ix[i]; v_055_Tm000 = dvislr[55].ix[i];
  v_056_Tm000 = dvislr[56].ix[i]; v_057_Tm000 = dvislr[57].ix[i]; v_058_Tm000 = dvislr[58].ix[i]; v_059_Tm000 = dvislr[59].ix[i]; v_060_Tm000 = dvislr[60].ix[i];

  v_061_Tm000 = dvislr[61].ix[i]; v_062_Tm000 = dvislr[62].ix[i]; v_063_Tm000 = dvislr[63].ix[i]; v_064_Tm000 = dvislr[64].ix[i]; v_065_Tm000 = dvislr[65].ix[i];
  v_066_Tm000 = dvislr[66].ix[i]; v_067_Tm000 = dvislr[67].ix[i]; v_068_Tm000 = dvislr[68].ix[i]; v_069_Tm000 = dvislr[69].ix[i]; v_070_Tm000 = dvislr[70].ix[i];

  v_071_Tm000 = dvislr[71].ix[i]; v_072_Tm000 = dvislr[72].ix[i]; v_073_Tm000 = dvislr[73].ix[i]; v_074_Tm000 = dvislr[74].ix[i]; v_075_Tm000 = dvislr[75].ix[i];
  v_076_Tm000 = dvislr[76].ix[i]; v_077_Tm000 = dvislr[77].ix[i]; v_078_Tm000 = dvislr[78].ix[i]; v_079_Tm000 = dvislr[79].ix[i]; v_080_Tm000 = dvislr[80].ix[i];

  v_081_Tm000 = dvislr[81].ix[i]; v_082_Tm000 = dvislr[82].ix[i]; v_083_Tm000 = dvislr[83].ix[i]; v_084_Tm000 = dvislr[84].ix[i]; v_085_Tm000 = dvislr[85].ix[i];
  v_086_Tm000 = dvislr[86].ix[i];

  # training_test_data: append data
  training_test_data = training_test_data.append(
    {
#       'usdmxn_logret_positive':usdmxn_logret_positive,
#       'usdmxn_logret_negative':usdmxn_logret_negative,
#       'usdmxn_logret_positive_0050bp':usdmxn_logret_positive_0050bp,
#       'usdmxn_logret_negative_0050bp':usdmxn_logret_negative_0050bp,
      
      'bc_up':bc_up,
      'bc_dn':bc_dn,
      
      '1':v_001_Tm000,  '2':v_002_Tm000, '3':v_003_Tm000, '4':v_004_Tm000, '5':v_005_Tm000,
      '6':v_006_Tm000,  '7':v_007_Tm000, '8':v_008_Tm000, '9':v_009_Tm000, '10':v_010_Tm000,

      '11':v_011_Tm000, '12':v_012_Tm000, '13':v_013_Tm000, '14':v_014_Tm000, '15':v_015_Tm000,
      '16':v_016_Tm000, '17':v_017_Tm000, '18':v_018_Tm000, '19':v_019_Tm000, '20':v_020_Tm000,

      '21':v_021_Tm000, '22':v_022_Tm000, '23':v_023_Tm000, '24':v_024_Tm000, '25':v_025_Tm000,
      '26':v_026_Tm000, '27':v_027_Tm000, '28':v_028_Tm000, '29':v_029_Tm000, '30':v_030_Tm000,

      '31':v_031_Tm000, '32':v_032_Tm000, '33':v_033_Tm000, '34':v_034_Tm000, '35':v_035_Tm000,
      '36':v_036_Tm000, '37':v_037_Tm000, '38':v_038_Tm000, '39':v_039_Tm000, '40':v_040_Tm000,

      '41':v_041_Tm000, '42':v_042_Tm000, '43':v_043_Tm000, '44':v_044_Tm000, '45':v_045_Tm000,
      '46':v_046_Tm000, '47':v_047_Tm000, '48':v_048_Tm000, '49':v_049_Tm000, '50':v_050_Tm000,

      '51':v_051_Tm000, '52':v_052_Tm000, '53':v_053_Tm000, '54':v_054_Tm000, '55':v_055_Tm000,
      '56':v_056_Tm000, '57':v_057_Tm000, '58':v_058_Tm000, '59':v_059_Tm000, '60':v_060_Tm000,

      '61':v_061_Tm000, '62':v_062_Tm000, '63':v_063_Tm000, '64':v_064_Tm000, '65':v_065_Tm000,
      '66':v_066_Tm000, '67':v_067_Tm000, '68':v_068_Tm000, '69':v_069_Tm000, '70':v_070_Tm000,

      '71':v_071_Tm000, '72':v_072_Tm000, '73':v_073_Tm000, '74':v_074_Tm000, '75':v_075_Tm000,
      '76':v_076_Tm000, '77':v_077_Tm000, '78':v_078_Tm000, '79':v_079_Tm000, '80':v_080_Tm000,
      
      '81':v_081_Tm000, '82':v_082_Tm000, '83':v_083_Tm000, '84':v_084_Tm000, '85':v_085_Tm000,
      '86':v_086_Tm000,
      
    },
    ignore_index=True)
  
# data_values_indexed_scaled_logret: row [7,4279] 
# training_test_data               : row [0,4272] = 4273 rows
# training_test_data: col 01-02 = 02 cols = binary outputs
# training_test_data: col 03-88 = 86 cols = inputs  


In [44]:
# training_test_data.describe()
training_test_data[['bc_up','bc_dn']].describe()

# check rows results
# dvislr_nr_m0010_z0000_p0010 rows =3554 - 7 = 3547
# dvislr_nr_m0020_z0000_p0020 rows =2904 - 7 = 2897
# dvislr_nr_m0030_z0000_p0030 rows =2316 - 7 = 2309
# dvislr_nr_m0040_z0000_p0040 rows =1837 - 7 = 1830
# dvislr_nr_m0050_z0000_p0050 rows =1437 - 7 = 1430
# dvislr_nr_m0060_z0000_p0060 rows =1135 - 7 = 1128
# dvislr_nr_m0070_z0000_p0070 rows =893
# dvislr_nr_m0080_z0000_p0080 rows =697
# dvislr_nr_m0090_z0000_p0090 rows =559
# dvislr_nr_m0100_z0000_p0100 rows =439

bc_up        bc_dn
count  1128.000000  1128.000000
mean      0.525709     0.474291
std       0.499560     0.499560
min       0.000000     0.000000
25%       0.000000     0.000000
50%       1.000000     0.000000
75%       1.000000     1.000000
max       1.000000     1.000000

### create the training and test data.

In [45]:
# 86 input columns
predictors_tf = training_test_data[training_test_data.columns[0+2+(31*2):0+2+(31*2)+86]]
predictors_tf.describe()

1            2            3            4            5  \
count  1128.000000  1128.000000  1128.000000  1128.000000  1128.000000   
mean      0.000921    -0.000612    -0.000107     0.000174     0.002763   
std       0.012126     0.027248     0.020950     0.004432     0.049899   
min      -0.066527    -0.614143    -0.101624    -0.020109    -0.837715   
25%      -0.008351    -0.008547    -0.011057    -0.002419    -0.005569   
50%       0.006223     0.000000     0.000000     0.000182     0.000000   
75%       0.009541     0.008484     0.009965     0.002867     0.010439   
max       0.070259     0.094618     0.130605     0.020071     0.700986   

                 6            7            8            9           10  \
count  1128.000000  1128.000000  1128.000000  1128.000000  1128.000000   
mean      0.000081    -0.000135     0.000507     0.000589     0.000264   
std       0.001005     0.014616     0.009634     0.037564     0.003739   
min      -0.006117    -0.146171    -0.053708    -0.385704    -0.024960   
25%      -0.000315    -0.000615    -0.002524    -0.007303    -0.001368   
50%       0.000046     0.000000     0.000000     0.000000     0.000197   
75%       0.000504     0.000731     0.002631     0.009637     0.001789   
max       0.005774     0.229973     0.082470     0.697777     0.035906   

          ...                77           78           79           80  \
count     ...       1128.000000  1128.000000  1128.000000  1128.000000   
mean      ...         -0.000825    -0.000780    -0.003498    -0.004030   
std       ...          0.020015     0.019885     0.110957     0.240191   
min       ...         -0.136928    -0.131822    -1.930917    -3.277145   
25%       ...         -0.010882    -0.010807    -0.020199    -0.024052   
50%       ...         -0.000875    -0.000371     0.000000     0.000000   
75%       ...          0.009629     0.009593     0.018318     0.021716   
max       ...          0.072736     0.082584     0.798726     4.235250   

                81           82           83           84           85  \
count  1128.000000  1128.000000  1128.000000  1128.000000  1128.000000   
mean     -0.009649    -0.000828    -0.000753    -0.000848    -0.000678   
std       0.378439     0.022573     0.021444     0.021239     0.021208   
min      -3.624341    -0.112705    -0.105936    -0.114583    -0.109933   
25%      -0.032944    -0.013403    -0.012643    -0.012278    -0.011888   
50%       0.000000    -0.001132    -0.000885    -0.000130    -0.001282   
75%       0.037170     0.012441     0.011955     0.011727     0.011705   
max       3.007205     0.081709     0.082212     0.078476     0.085061   

                86  
count  1128.000000  
mean      0.000185  
std       0.005680  
min      -0.013134  
25%      -0.000433  
50%       0.000000  
75%       0.000586  
max       0.176596  

[8 rows x 86 columns]

In [46]:
# 2 output columns
classes_tf = training_test_data[training_test_data.columns[0:2]] # col 0, 1
# classes_tf = training_test_data[training_test_data.columns[2:4]] # col 2, 3
classes_tf.describe()

bc_up        bc_dn
count  1128.000000  1128.000000
mean      0.525709     0.474291
std       0.499560     0.499560
min       0.000000     0.000000
25%       0.000000     0.000000
50%       1.000000     0.000000
75%       1.000000     1.000000
max       1.000000     1.000000

In [47]:
# Split: train data = 80%
training_set_size = int(len(training_test_data) * 0.8)
print 'training_set_size=' + str(training_set_size )
# Split: test data = 20%
test_set_size = len(training_test_data) - training_set_size
print 'test_set_size=' + str(test_set_size )

# After noise removed
# training_set_size=1144
# test_set_size=286

# noise remove stratification: [-0010,+0010]: training_set_size=2837, test_set_size=710
# noise remove stratification: [-0020,+0020]: training_set_size=2317, test_set_size=580
# noise remove stratification: [-0030,+0030]: training_set_size=1847, test_set_size=462
# noise remove stratification: [-0040,+0040]: training_set_size=1464, test_set_size=366
# noise remove stratification: [-0050,+0050]: training_set_size=1144, test_set_size=286
# noise remove stratification: [-0060,+0060]: training_set_size=902,  test_set_size=226
# noise remove stratification: [-0070,+0070]: 
# noise remove stratification: [-0080,+0080]: 
# noise remove stratification: [-0090,+0090]: 
# noise remove stratification: [-0100,+0100]: 



training_set_size=902
test_set_size=226


## Split into training and test data

In [48]:
training_predictors_tf = predictors_tf[:training_set_size]
training_classes_tf = classes_tf[:training_set_size]
test_predictors_tf = predictors_tf[training_set_size:]
test_classes_tf = classes_tf[training_set_size:]


Define some metrics here to evaluate the models.

* [Precision](https://en.wikipedia.org/wiki/Precision_and_recall#Precision) -  The ability of the classifier not to label as positive a sample that is negative.
* [Recall](https://en.wikipedia.org/wiki/Precision_and_recall#Recall) - The ability of the classifier to find all the positive samples.
* [F1 Score](https://en.wikipedia.org/wiki/F1_score) - A weighted average of the precision and recall, where an F1 score reaches its best value at 1 and worst score at 0.
* Accuracy - The percentage correctly predicted in the test data.

In [49]:
def tf_confusion_metrics(model, actual_classes, session, feed_dict):
  predictions = tf.argmax(model, 1)
  actuals = tf.argmax(actual_classes, 1)

  ones_like_actuals = tf.ones_like(actuals)
  zeros_like_actuals = tf.zeros_like(actuals)
  ones_like_predictions = tf.ones_like(predictions)
  zeros_like_predictions = tf.zeros_like(predictions)

  tp_op = tf.reduce_sum(
    tf.cast(
      tf.logical_and(
        tf.equal(actuals, ones_like_actuals), 
        tf.equal(predictions, ones_like_predictions)
      ), 
      "float"
    )
  )

  tn_op = tf.reduce_sum(
    tf.cast(
      tf.logical_and(
        tf.equal(actuals, zeros_like_actuals), 
        tf.equal(predictions, zeros_like_predictions)
      ), 
      "float"
    )
  )

  fp_op = tf.reduce_sum(
    tf.cast(
      tf.logical_and(
        tf.equal(actuals, zeros_like_actuals), 
        tf.equal(predictions, ones_like_predictions)
      ), 
      "float"
    )
  )

  fn_op = tf.reduce_sum(
    tf.cast(
      tf.logical_and(
        tf.equal(actuals, ones_like_actuals), 
        tf.equal(predictions, zeros_like_predictions)
      ), 
      "float"
    )
  )

  tp, tn, fp, fn = \
    session.run(
      [tp_op, tn_op, fp_op, fn_op], 
      feed_dict
    )

  tpr = float(tp)/(float(tp) + float(fn))
  fpr = float(fp)/(float(tp) + float(fn))

  accuracy = (float(tp) + float(tn))/(float(tp) + float(fp) + float(fn) + float(tn))

  recall = tpr
  precision = float(tp)/(float(tp) + float(fp))
  
  f1_score = (2 * (precision * recall)) / (precision + recall)
  
  print 'Precision = ', precision
  print 'Recall = ', recall
  print 'F1 Score = ', f1_score
  print 'Accuracy = ', accuracy

### Binary classification with TensorFlow

Now, get some tensors flowing. The model is binary classification expressed in TensorFlow.

In [50]:
# tensorflow session
sess = tf.Session()

In [51]:
# Define variables for the number of predictors and number of classes to remove magic numbers from our code.
num_predictors = len(training_predictors_tf.columns) # 24-6=18 in the default case
num_classes = len(training_classes_tf.columns) # 2 in the default case
print 'num_predictors=' + str(num_predictors)
print 'num_classes=' + str(num_classes)

num_predictors=86
num_classes=2


In [52]:
# Define placeholders for the data we feed into the process - feature data and actual classes.
# input = 18
feature_data = tf.placeholder("float", [None, num_predictors])
# output = 2
actual_classes = tf.placeholder("float", [None, num_classes])

In [53]:
# Define a matrix of weights and initialize it with some small random values.
weights = tf.Variable(tf.truncated_normal([num_predictors, num_classes], stddev=0.0001))
biases = tf.Variable(tf.ones([num_classes]))

In [54]:
# Define our model...
# Here we take a softmax regression of the product of our feature data and weights.
model = tf.nn.softmax(tf.matmul(feature_data, weights) + biases)

# Define a cost function (we're using the cross entropy).
cost = -tf.reduce_sum(actual_classes*tf.log(model))

# Define a training step...
# Here we use gradient descent with a learning rate of 0.01 using the cost function we just defined.
training_step = tf.train.AdamOptimizer(learning_rate=0.0001).minimize(cost)

init = tf.initialize_all_variables()
sess.run(init)

In [55]:
# BEFORE the model has been run, ie. not yet trained
# display weights
w_86_2 = sess.run(weights)    
print w_86_2 

# Expect something like this (small numbers e-05): 
# [[ -6.66152628e-05  -6.49469657e-05]
#  [  2.86067752e-05   3.27789494e-05]
#  [  4.73170294e-05   1.85125769e-04]
#  [ -6.42955492e-05   4.16024632e-05]
#  [ -1.54488771e-05  -2.10903818e-05]
#  [  4.62506796e-05  -2.98624745e-05]
#  [ -5.35508079e-05  -1.30856875e-04]
#  [ -1.70812025e-04   1.33106529e-04]
#  [ -1.45097118e-04  -1.80459567e-04]
#  [  1.02448161e-04   8.27739277e-05]
#  [ -1.33277281e-04  -4.04360726e-05]
#  [ -1.36186543e-04   8.62382149e-05]
#  [  2.80324894e-05  -2.20580205e-05]
#  [ -4.52000713e-05   2.54859442e-05]
#  [ -6.20259525e-05  -6.95227573e-05]
#  [  1.68935469e-04  -4.03221093e-05]
#  [ -9.51008842e-05   4.10227585e-05]
#  [ -5.90909331e-05  -9.20566745e-05]]

[[ -1.19075463e-04   5.21266084e-05]
 [ -7.93897107e-06  -3.01031469e-06]
 [  4.90011917e-05  -5.81187051e-05]
 [  1.21333414e-05   1.45631580e-04]
 [ -1.32475485e-04  -9.35834178e-05]
 [  2.40627087e-05   4.88963524e-05]
 [  5.46589181e-05  -9.34279742e-05]
 [  8.56469342e-06   1.75672685e-04]
 [ -9.23640400e-05  -7.35338617e-05]
 [ -8.66797054e-05   9.84547223e-05]
 [ -5.46369556e-05  -1.41040684e-04]
 [  4.45059231e-05   3.50342307e-05]
 [ -2.30454339e-06   1.12168687e-04]
 [  2.35478656e-05  -9.92986679e-05]
 [ -9.31521645e-05  -4.61520722e-05]
 [  2.63135062e-05   9.50964313e-05]
 [ -1.53614237e-04   7.12975161e-05]
 [ -3.20473919e-05   3.06708425e-05]
 [  6.16221441e-05  -2.55244577e-05]
 [ -2.03459294e-05  -3.29161376e-05]
 [ -1.84350618e-04  -6.06808771e-06]
 [ -6.34520911e-05  -8.38328288e-06]
 [ -7.84539880e-05  -3.64078551e-05]
 [ -2.09466943e-05  -1.86818317e-04]
 [  1.44720660e-04  -1.38785632e-04]
 [  1.41966433e-04  -4.01814432e-05]
 [ -7.50212275e-05   1.12914917e-04]
 

We'll train our model in the following snippet. The approach of TensorFlow to executing graph operations allows fine-grained control over the process. Any operation you provide to the session as part of the run operation will be executed and the results returned. You can provide a list of multiple operations.

You'll train the model over 30,000 iterations using the full dataset each time. Every thousandth iteration we'll assess the accuracy of the model on the training data to assess progress.

In [56]:
correct_prediction = tf.equal(tf.argmax(model, 1), tf.argmax(actual_classes, 1))
accuracy = tf.reduce_mean(tf.cast(correct_prediction, "float"))

for i in range(1, 30001):
  sess.run(
    training_step, 
    feed_dict={
      feature_data: training_predictors_tf.values, 
      actual_classes: training_classes_tf.values.reshape(len(training_classes_tf.values), 2)
    }
  )
  if i%5000 == 0:
    print i, sess.run(
      accuracy,
      feed_dict={
        feature_data: training_predictors_tf.values, 
        actual_classes: training_classes_tf.values.reshape(len(training_classes_tf.values), 2)
      }
    )
    
# Expect:
# 5000 0.603862
# 10000 0.630778
# 15000 0.639555
# 20000 0.654184
# 25000 0.660035
# 30000 0.666764

# After using < -0.5% and > +0.5%
# 5000 0.928906
# 10000 0.928906
# 15000 0.928906
# 20000 0.928906
# 25000 0.928906
# 30000 0.928906

# Noise Removal: [-0010, +0010]
# 5000 0.617201
# 10000 0.64399
# 15000 0.665139
# 20000 0.683468
# 25000 0.688403
# 30000 0.691928

# Noise Removal: [-0020, +0020]
# 5000 0.625809
# 10000 0.6612
# 15000 0.680622
# 20000 0.695727
# 25000 0.708675
# 30000 0.712128

# Noise Removal: [-0030, +0030]
# 5000 0.635084
# 10000 0.677315
# 15000 0.701678
# 20000 0.714672
# 25000 0.729291
# 30000 0.738495

# Noise Removal: [-0040, +0040]
# 5000 0.645492
# 10000 0.689891
# 15000 0.717213
# 20000 0.734973
# 25000 0.748634
# 30000 0.758197

# Noise Removal: [-0050, +0050]
# 5000 0.653846
# 10000 0.702797
# 15000 0.730769
# 20000 0.740385
# 25000 0.75
# 30000 0.768357

# Noise Removal: [-0060, +0060]
# 5000 0.672949
# 10000 0.72173
# 15000 0.750554
# 20000 0.761641
# 25000 0.783814
# 30000 0.792683

# Noise Removal: [-0070, +0070]

# Noise Removal: [-0080, +0080]

# Noise Removal: [-0090, +0090]

# Noise Removal: [-0100, +0100]


5000 0.672949
10000 0.72173
15000 0.750554
20000 0.761641
25000 0.783814
30000 0.792683


Using all data:
Accuracy  66.6% on training data
That is OK, better than random.

After noise removed, We get 77 % accuracy on training data.

## Accuracy on test data

In [57]:
feed_dict= {
  feature_data: test_predictors_tf.values,
  actual_classes: test_classes_tf.values.reshape(len(test_classes_tf.values), 2)
}

tf_confusion_metrics(model, actual_classes, sess, feed_dict)

# Precision =  0.541176470588
# Recall =  0.555555555556
# F1 Score =  0.548271752086
# Accuracy =  0.556725146199

# After removing noise
# Precision =  0.684210526316
# Recall =  0.534246575342
# F1 Score =  0.6
# Accuracy =  0.636363636364

# Noise Removal: [-0010, +0010]
# Precision =  0.577235772358
# Recall =  0.605113636364
# F1 Score =  0.590846047157
# Accuracy =  0.584507042254

# Noise Removal: [-0020, +0020]
# Precision =  0.588850174216
# Recall =  0.603571428571
# F1 Score =  0.596119929453
# Accuracy =  0.605172413793

# Noise Removal: [-0030, +0030]
# Precision =  0.616740088106
# Recall =  0.619469026549
# F1 Score =  0.618101545254
# Accuracy =  0.625541125541

# Noise Removal: [-0040, +0040]
# Precision =  0.622093023256
# Recall =  0.58152173913
# F1 Score =  0.601123595506
# Accuracy =  0.612021857923

# Noise Removal: [-0050, +0050]
# Precision =  0.684210526316
# Recall =  0.534246575342
# F1 Score =  0.6
# Accuracy =  0.636363636364

# Noise Removal: [-0060, +0060]
# Precision =  0.644736842105
# Recall =  0.453703703704
# F1 Score =  0.532608695652
# Accuracy =  0.619469026549

# Noise Removal: [-0070, +0070]

# Noise Removal: [-0080, +0080]

# Noise Removal: [-0090, +0090]

# Noise Removal: [-0100, +0100]


Precision =  0.644736842105
Recall =  0.453703703704
F1 Score =  0.532608695652
Accuracy =  0.619469026549


Using all data, Accuracy 55.6% on test data.
After removing noise, Accuracy 64% on test data.

In [58]:
# AFTER the model has been run, ie. trained
# display weights
w_86_2 = sess.run(weights)    
print w_86_2 

[[ 2.96132183 -2.96138859]
 [-2.81565118  2.81564021]
 [-2.91209364  2.91208458]
 [-1.96235788  1.96251559]
 [ 2.8368926  -2.83711863]
 [ 2.75949311 -2.75942016]
 [-1.98157299  1.98153412]
 [ 2.53368449 -2.53349996]
 [ 2.89654422 -2.89671016]
 [-0.23813249  0.23814453]
 [-2.81490278  2.81470704]
 [-2.73814321  2.73822236]
 [ 0.22130398 -0.22119404]
 [-2.75612378  2.7560482 ]
 [-0.17750882  0.17736995]
 [-1.80428767  1.80440652]
 [ 0.07182407 -0.0720861 ]
 [-0.07663839  0.07702472]
 [-0.00562787  0.00448106]
 [ 0.93367016 -0.93386751]
 [-0.08873752  0.08757472]
 [-2.79255915  2.79248691]
 [-0.00924385  0.00885756]
 [-1.18252063  1.1822232 ]
 [ 2.76867199 -2.76866579]
 [-2.60087967  2.60098171]
 [-2.60856295  2.60860109]
 [-2.54400396  2.54422426]
 [-0.5156337   0.51574874]
 [-0.04585199  0.0454029 ]
 [-0.33829668  0.33898139]
 [-0.3114343   0.31094575]
 [-0.27259123  0.27281645]
 [ 0.10814866 -0.10801176]
 [ 3.05603099 -3.05604649]
 [ 0.65164852 -0.65340197]
 [-2.87353039  2.87359715]
 

In [126]:
# grab a sample input row from the test set
sample_input_1_86 = test_predictors_tf[:10]
# HTML(pd.DataFrame(sample_input_1_86).to_html())
sample_input_1_86

1         2         3         4    5         6    7         8  \
2837  0.001287  0.025642  0.036871 -0.006917  0.0 -0.001215  0.0 -0.001086   
2838 -0.004996  0.074108  0.030992 -0.008273  0.0 -0.001400  0.0 -0.003413   
2839 -0.011748  0.000000 -0.004831  0.007302  0.0  0.000746  0.0  0.000000   
2840  0.008148  0.000000 -0.004808  0.002475  0.0 -0.000373  0.0  0.000000   
2841 -0.008101  0.000000  0.000000 -0.001769  0.0 -0.000187  0.0 -0.001342   
2842 -0.006687  0.000000  0.007290  0.005704  0.0  0.000280  0.0 -0.001211   
2843  0.017320  0.024780 -0.007290  0.002685  0.0  0.000374  0.0 -0.002059   
2844 -0.002766  0.054505  0.000000 -0.000448  0.0  0.000561  0.0 -0.002837   
2845  0.027181  0.000000  0.000000  0.001345  0.0 -0.000935  0.0 -0.001302   
2846  0.004090 -0.065201 -0.014423  0.007112  0.0  0.002058  0.0  0.002474   

        9        10 ...         77        78        79        80        81  \
2837  0.0 -0.006211 ...   0.020790  0.022198  0.014982  0.014815  0.024571   
2838  0.0  0.000298 ...   0.011130  0.007887  0.003781  0.000000 -0.014815   
2839  0.0 -0.001342 ...  -0.014738 -0.013037  0.015267  0.015038  0.024815   
2840  0.0 -0.003274 ...   0.002930  0.002236  0.007722  0.000000  0.005038   
2841  0.0  0.000149 ...  -0.010107 -0.010464  0.007782  0.007605  0.000000   
2842  0.0  0.000000 ...  -0.010863 -0.010792  0.007722  0.007547  0.004963   
2843  0.0  0.018012 ...   0.008681  0.007110 -0.015385 -0.022473 -0.019705   
2844  0.0  0.000761 ...  -0.005868 -0.005391 -0.018904 -0.011050 -0.014528   
2845  0.0  0.001828 ...   0.010236  0.010159  0.011300  0.022223  0.009662   
2846  0.0  0.005198 ...  -0.025290 -0.024677  0.000000 -0.003738  0.009756   

            82        83        84        85   86  
2837  0.023684  0.020283  0.017523  0.018177  0.0  
2838  0.012799  0.012440  0.012383  0.010842  0.0  
2839 -0.009982 -0.008310 -0.007673 -0.006944  0.0  
2840 -0.010126 -0.010860 -0.010956 -0.010878  0.0  
2841  0.003648  0.003075  0.002895  0.004204  0.0  
2842 -0.017285 -0.013983 -0.012811 -0.008835  0.0  
2843 -0.002345 -0.001982 -0.003231 -0.005776  0.0  
2844  0.010121  0.008394  0.008639  0.005240  0.0  
2845 -0.005426 -0.002658 -0.002167  0.001824  0.0  
2846 -0.016336 -0.015477 -0.012690 -0.012489  0.0  

[10 rows x 86 columns]

In [127]:
sample_input_1_86.describe()

1          2          3          4     5          6     7  \
count  10.000000  10.000000  10.000000  10.000000  10.0  10.000000  10.0   
mean    0.002373   0.011384   0.004380   0.000922   0.0  -0.000009   0.0   
std     0.012239   0.037550   0.016630   0.005403   0.0   0.001044   0.0   
min    -0.011748  -0.065201  -0.014423  -0.008273   0.0  -0.001400   0.0   
25%    -0.006265   0.000000  -0.004825  -0.001438   0.0  -0.000794   0.0   
50%    -0.000739   0.000000   0.000000   0.001910   0.0   0.000047   0.0   
75%     0.007134   0.025427   0.005468   0.004949   0.0   0.000514   0.0   
max     0.027181   0.074108   0.036871   0.007302   0.0   0.002058   0.0   

               8     9         10  ...          77         78         79  \
count  10.000000  10.0  10.000000  ...   10.000000  10.000000  10.000000   
mean   -0.001078   0.0   0.001542  ...   -0.001310  -0.001477   0.003427   
std     0.001653   0.0   0.006522  ...    0.014295   0.013848   0.011813   
min    -0.003413   0.0  -0.006211  ...   -0.025290  -0.024677  -0.018904   
25%    -0.001880   0.0  -0.001007  ...   -0.010674  -0.010710   0.000945   
50%    -0.001256   0.0   0.000224  ...   -0.001469  -0.001578   0.007722   
75%    -0.000271   0.0   0.001561  ...    0.009847   0.007693   0.010420   
max     0.002474   0.0   0.018012  ...    0.020790   0.022198   0.015267   

              80         81         82         83         84         85    86  
count  10.000000  10.000000  10.000000  10.000000  10.000000  10.000000  10.0  
mean    0.002997   0.002976  -0.001125  -0.000908  -0.000809  -0.000463   0.0  
std     0.013357   0.015593   0.013460   0.011883   0.010832   0.010156   0.0  
min    -0.022473  -0.019705  -0.017285  -0.015477  -0.012811  -0.012489   0.0  
25%    -0.002804  -0.010896  -0.010090  -0.010223  -0.010135  -0.008362   0.0  
50%     0.003774   0.005000  -0.003886  -0.002320  -0.002699  -0.001976   0.0  
75%     0.013012   0.009733   0.008503   0.007064   0.007203   0.004981   0.0  
max     0.022223   0.024815   0.023684   0.020283   0.017523   0.018177   0.0  

[8 rows x 86 columns]

In [128]:
# use the weights from our model to make a prediction
output_1_2 = np.dot(sample_input_1_86, w_86_2) > 0
print output_1_2
usdmxn_up = output_1_2[0,0]
usdmxn_dn = output_1_2[0,1]
if usdmxn_up >= usdmxn_dn:
  usdmxn_pred = 1 # up
else:
  usdmxn_pred = -1 # down
print 'usdmxn_pred = '+ str(usdmxn_pred)

[[ True False]
 [False  True]
 [ True False]
 [ True False]
 [False  True]
 [False  True]
 [ True False]
 [False  True]
 [ True False]
 [ True False]]
usdmxn_pred = 1


# Model 2: Neural Network (NN): Feed-Forward with 2 Hidden Layers

You'll now build a proper feed-forward neural net with two hidden layers.

In [ ]:
# tensorflow session new
sess1 = tf.Session()

In [ ]:
# inputs
num_predictors = len(training_predictors_tf.columns)
print 'num_predictors=' + str(num_predictors)

# outputs
num_classes = len(training_classes_tf.columns)
print 'num_classes=' + str(num_classes)

In [ ]:
feature_data = tf.placeholder("float", [None, num_predictors])
actual_classes = tf.placeholder("float", [None, 2])

In [ ]:
# layer 1
i_01_in = 86
i_01_out = 100
weights1 = tf.Variable(tf.truncated_normal([i_01_in, i_01_out], stddev=0.0001))
biases1 = tf.Variable(tf.ones([i_01_out]))

In [ ]:
# layer 2
i_02_in = i_01_out
i_02_out = 25
weights2 = tf.Variable(tf.truncated_normal([i_02_in, i_02_out], stddev=0.0001))
biases2 = tf.Variable(tf.ones([i_02_out]))

In [ ]:
# layer 3
i_03_in = i_02_out
i_03_out = 2
weights3 = tf.Variable(tf.truncated_normal([i_03_in, i_03_out], stddev=0.0001))
biases3 = tf.Variable(tf.ones([i_03_out]))

In [ ]:
# neural net
hidden_layer_0 = feature_data
hidden_layer_1 = tf.nn.relu(tf.matmul(hidden_layer_0, weights1) + biases1)
hidden_layer_2 = tf.nn.relu(tf.matmul(hidden_layer_1, weights2) + biases2)
hidden_layer_3 = tf.matmul(hidden_layer_2, weights3) + biases3

model = tf.nn.softmax(hidden_layer_3)

cost = -tf.reduce_sum(actual_classes*tf.log(model))

train_op1 = tf.train.AdamOptimizer(learning_rate=0.0001).minimize(cost)

init = tf.initialize_all_variables()
sess1.run(init)

In [ ]:
# BEFORE the model has been run, ie. trained
nn_e1 = sess1.run(biases1)    
# HTML(pd.DataFrame(nn_e1).transpose().to_html())
pd.DataFrame(nn_e1).transpose()


In [ ]:
# BEFORE the model has been run, ie. trained
nn_w1 = sess1.run(weights1)    
# HTML(pd.DataFrame(nn_w1).to_html())
pd.DataFrame(nn_w1)


In [ ]:
# BEFORE the model has been run, ie. trained
nn_e2 = sess1.run(biases2)    
# HTML(pd.DataFrame(nn_e2).transpose().to_html())
pd.DataFrame(nn_e2).transpose()


In [ ]:
# BEFORE the model has been run, ie. trained
# display weights
nn_w2 = sess1.run(weights2)    
# HTML(pd.DataFrame(nn_w2).to_html())
pd.DataFrame(nn_w2)


In [ ]:
# BEFORE the model has been run, ie. trained
nn_e3 = sess1.run(biases3)    
# HTML(pd.DataFrame(nn_e3).transpose().to_html())
pd.DataFrame(nn_e3).transpose()


In [ ]:
# BEFORE the model has been run, ie. trained
# display weights
nn_w3 = sess1.run(weights3)    
# HTML(pd.DataFrame(nn_w3).to_html())
pd.DataFrame(nn_w3)


Again, you'll train the model over 10,000 iterations using the full dataset each time. Every thousandth iteration, you'll assess the accuracy of the model on the training data to assess progress.

In [ ]:
correct_prediction = tf.equal(tf.argmax(model, 1), tf.argmax(actual_classes, 1))
accuracy = tf.reduce_mean(tf.cast(correct_prediction, "float"))
num_iterations = 10001
num_iterations_until_diaplay = 1000

for i in range(1, num_iterations):
  sess1.run(
    train_op1, 
    feed_dict={
      feature_data: training_predictors_tf.values, 
      actual_classes: training_classes_tf.values.reshape(len(training_classes_tf.values), 2)
    }
  )
  if i%num_iterations_until_diaplay == 0:
    print i, sess1.run(
      accuracy,
      feed_dict={
        feature_data: training_predictors_tf.values, 
        actual_classes: training_classes_tf.values.reshape(len(training_classes_tf.values), 2)
      }
    )
  
  
  
# 1000 0.67086
# 2000 0.765652
# 3000 0.828847
# 4000 0.890287
# 5000 0.937975
# 6000 0.956115
# 7000 0.971913
# 8000 0.979813
# 9000 0.985079
# 10000 0.987712

# 5000 0.93505
# 10000 0.988297
# 15000 0.491223 # overfit?
# 20000 0.491223
# 25000 0.491223
# 30000 0.491223


A significant improvement in accuracy with the training data shows that the hidden layers are adding additional capacity for learning to the model.

Looking at precision, recall, and accuracy, you can see a measurable improvement in performance, but certainly not a [step function](https://wikipedia.org/wiki/Step_function). This indicates that we're likely reaching the limits of this relatively simple feature set.

In [ ]:
feed_dict= {
  feature_data: test_predictors_tf.values,
  actual_classes: test_classes_tf.values.reshape(len(test_classes_tf.values), 2)
}

tf_confusion_metrics(model, actual_classes, sess1, feed_dict)

# Precision =  0.936768149883
# Recall =  0.966183574879
# F1 Score =  0.951248513674
# Accuracy =  0.952046783626


In [ ]:
# AFTER the model has been run, ie. trained
nn_e1 = sess1.run(biases1)    
# HTML(pd.DataFrame(nn_e1).transpose().to_html())
pd.DataFrame(nn_e1).transpose()


In [ ]:
# AFTER the model has been run, ie. trained
# display weights
nn_w1 = sess1.run(weights1)    
# HTML(pd.DataFrame(nn_w1).to_html())
pd.DataFrame(nn_w1)


In [ ]:
# AFTER the model has been run, ie. trained
nn_e2 = sess1.run(biases2)    
# HTML(pd.DataFrame(nn_e2).transpose().to_html())
pd.DataFrame(nn_e2).transpose()


In [ ]:
# AFTER the model has been run, ie. trained
# display weights
nn_w2 = sess1.run(weights2)    
# HTML(pd.DataFrame(nn_w2).to_html())
pd.DataFrame(nn_w2)


In [ ]:
# AFTER the model has been run, ie. trained
nn_e3 = sess1.run(biases3)    
# HTML(pd.DataFrame(nn_e3).transpose().to_html())
pd.DataFrame(nn_e3).transpose()


In [ ]:
# AFTER the model has been run, ie. trained
# display weights
nn_w3 = sess1.run(weights3)    
# HTML(pd.DataFrame(nn_w3).to_html())
pd.DataFrame(nn_w3)


In [ ]:
# # AFTER the model has been run, ie. trained
# nn_hl3 = sess1.run(hidden_layer_3)    
# print nn_hl3
# # HTML(pd.DataFrame(nn_hl3).to_html())
# pd.DataFrame(nn_hl3)


## Conclusion

You've covered a lot of ground. You moved from sourcing five years of financial time-series data, to munging that data into a more suitable form. You explored and visualized that data with exploratory data analysis and then decided on a machine learning model and the features for that model. You engineered those features, built a binary classifier in TensorFlow, and analyzed its performance. You built a feed forward neural net with two hidden layers in TensorFlow and analyzed its performance.

How did the technology fare? It should take most people 1.5 to 3 hours to extract the juice from this solution, and none of that time is spent waiting for infrastructure or software; it's spent reading and thinking. In many organizations, it can take anywhere from days to months to do this sort of data analysis, depending on whether you need to procure any hardware. And you didn't need to do anything with infrastructure or additional software. Rather, you used a web-based console to direct GCP to set up systems on your behalf, which it did—fully managed, maintained, and supported—freeing you up to spend your time analyzing. 

It was also cost effective. If you took your time with this solution and spent three hours to go through it, the cost would be a few pennies. 

Cloud Datalab worked admirably, too. iPython/Jupyter has always been a great platform for interactive, iterative work and a fully-managed version of that platform on GCP, with connectors to other GCP technologies such as BigQuery and Google Cloud Storage, is a force multiplier for your analysis needs.  If you haven't used iPython before, this solution might have been eye opening, for you. If you're already familiar with iPython, then you'll love the connectors to other GCP technologies.

Of course, R and Matlab are popular tools in machine learning, and we've made no mention either in this solution. Neither R nor Matlab are available as managed services on GCP. Both can be hosted in GCP and accessed through a cloud-friendly, web frontend.

TensorFlow is a special piece of technology. It is expressive, performs well, and comes with the weight of Google's machine learning history and expertise to back it up and support it. We've only scratched the surface, but you can already see that within a handful of lines of code we've been able to write two models. Neither of them is cutting edge, by design, but neither of them is trivial either. With some additional tuning they would suit a whole spectrum of machine learning tasks. 

Finally, how did we do with the data analysis? We did well: over 70% accuracy in predicting the close of the S&P 500 is the highest we've seen achieved on this dataset, so with few steps and a few lines of code we've produced a full-on machine learning model. The reason for the relatively modest accuracy achieved is the dataset itself; there isn't enough signal there to do significantly better. But 7 times out of 10, we were able to correctly determine if the S&P 500 index would close up or down on the day, and that's objectively good.

## Cleanup

**When you're finished, shut down the managed VM you used for Cloud Datalab to avoid incurring costs.**